### PR TITLE
[Finder] Optimize the hot-path

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #15824 |
| License | MIT |
| Doc PR | - |

A significant part of the perf gain in #15802 was related to filters not being applied recursively...
#15824 fixing this, performance dropped again.

This PR optimizes the hot path by replacing a regexp test by a simple `isset` when possible.
Blackfire diff after #15824 is the following:
https://blackfire.io/profiles/compare/9e489018-998d-4acb-92a0-46011828e83b/graph

`preg_match` is not called anymore, and  `Symfony\Component\Finder\Iterator\RecursiveDirectoryIterator::current()` is also cut by two.

When this `isset` optimization is disabled and replaced by a concatenation of all the regexps patterns in a single bigger one, the gain is still significant but lower:
https://blackfire.io/profiles/compare/db86b80e-b63e-4fc9-9ff3-9ed32baeb948/graph

This makes me think that an other and last round of optimization is possible by merging all regexps in one in MultiplePcreFilterIterator filters. If someone wants to work on this, please do it :)

### CODE REVIEW

**Comment by stof** [comment]
> This makes me think that an other and last round of optimization is possible by merging all regexps in one in MultiplePcreFilterIterator filters. If someone wants to work on this, please do it :)

this is quite hard to do though, as it is not just a matter of concatenating regexes (combining regexes is very hard to do, and almost impossible in case complex regexes are involved)

File: src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php [hunk 1]
```diff
@@ -73,6 +87,9 @@ public function getChildren()
             if ($children instanceof self) {
                 // parent method will call the constructor with default arguments, so unreadable dirs won't be ignored anymore
                 $children->ignoreUnreadableDirs = $this->ignoreUnreadableDirs;
+                $children->subPath = $children->getSubPath();
```
Comment (stof) [inline src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php]: Please add a command before this to explain it is a performance optimization to avoid redoing the same work in all childrens

**Comment by stof** [comment]
@nicolas-grekas this seems to make PHP segfault

**Comment by stof** [comment]
@nicolas-grekas can you post the profiling and benchmarking results of the ExcludeDirectoryFilter optimization you posted, as this is a totally different optimization than the initial one ?

File: src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php [hunk 2]
```diff
@@ -31,6 +31,10 @@ class RecursiveDirectoryIterator extends \RecursiveDirectoryIterator
      */
     private $rewindable;
 
+    private $rootPath;
+    private $subPath;
+    private $directorySeparator = '/';
```
Comment (stof) [inline src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php]: please add some documentation on them to make the code understandable by someone reading the code later

**Comment by nicolas-grekas** [comment]
@stof the profiling of the ExcludeDirectoryFilter is embedded in the above profiles: you need to click on the "current" node (bottom right) then on the magnifying glass on the left box to focus draw the graph around this node.

File: src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php [hunk 3]
```diff
@@ -31,6 +31,11 @@ class RecursiveDirectoryIterator extends \RecursiveDirectoryIterator
      */
     private $rewindable;
 
+    // these 3 properties take part of the performance optimization to avoid redoing the same work in all iterations
+    private $rootPath;
+    private $subPath = false;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php]: null is the return value of $this->getSubPath() on hhvm, thus requiring `false` here as "not set" placeholder

**Comment by nicolas-grekas** [comment]
```
   1010 files
    110 directories
     10 iterations
      6 cases
      2 adapters

            case             php        gnu_find
               0          92,396          33,565
               1          99,673          32,776
               2         103,553          54,970
               3         102,736          61,805
               4         105,615          19,072
               5         107,359          35,432

      0 Find files by name containing a*
      1 Find files by name containing a* not containing *a
      2 Find files by name containing ~^a.*~ not containing ~.*a$~
      3 Find files by path containing a*
      4 Find files by path containing a* not containing *a
      5 Find files by path containing ~^a.*~ not containing ~.*a$~
```
See https://github.com/symfony/symfony/pull/15802#issuecomment-140331104 for comparison.

**Comment by nicolas-grekas** [comment]
I managed merging several regexps together using the following path, but I my first quick tests didn't prove much benefit. If someone else wanna try...

```diff
diff --git a/src/Symfony/Component/Finder/Expression/Regex.php b/src/Symfony/Component/Finder/Expression/Regex.php
index a249fc2..cd232f9 100644
--- a/src/Symfony/Component/Finder/Expression/Regex.php
+++ b/src/Symfony/Component/Finder/Expression/Regex.php
@@ -61,7 +61,7 @@ class Regex implements ValueInterface
      */
     public static function create($expr)
     {
-        if (preg_match('/^(.{3,}?)([imsxuADU]*)$/', $expr, $m)) {
+        if (preg_match('/^(.{2,}?)([imsxuADU]*)$/', $expr, $m)) {
             $start = substr($m[1], 0, 1);
             $end = substr($m[1], -1);
 
diff --git a/src/Symfony/Component/Finder/Iterator/MultiplePcreFilterIterator.php b/src/Symfony/Component/Finder/Iterator/MultiplePcreFilterIterator.php
index d21380d..ec1fd35 100644
--- a/src/Symfony/Component/Finder/Iterator/MultiplePcreFilterIterator.php
+++ b/src/Symfony/Component/Finder/Iterator/MultiplePcreFilterIterator.php
@@ -12,6 +12,7 @@
 namespace Symfony\Component\Finder\Iterator;
 
 use Symfony\Component\Finder\Expression\Expression;
+use Symfony\Component\Finder\Expression\Regex;
 
 /**
  * MultiplePcreFilterIterator filters files using patterns (regexps, globs or strings).
@@ -32,14 +33,34 @@ abstract class MultiplePcreFilterIterator extends FilterIterator
      */
     public function __construct(\Iterator $iterator, array $matchPatterns, array $noMatchPatterns)
     {
+        $regexps = array();
         $this->matchRegexps = array();
         foreach ($matchPatterns as $pattern) {
-            $this->matchRegexps[] = $this->toRegex($pattern);
+            $regex = Regex::create($this->toRegex($pattern))->render();
+
+            if (preg_match('/\W([imsxUXJ]*)$/', $regex, $match)) {
+                $regexps[] = '(?'.$match[1].':'.substr($regex, 1, -strlen($match[0])).')';
+            } else {
+                $this->matchRegexps[] = $regex;
+            }
+        }
+        if ($regexps) {
+            $this->matchRegexps[] = Regex::BOUNDARY.implode('|', $regexps).Regex::BOUNDARY;
         }
 
+        $regexps = array();
         $this->noMatchRegexps = array();
         foreach ($noMatchPatterns as $pattern) {
-            $this->noMatchRegexps[] = $this->toRegex($pattern);
+            $regex = Regex::create($this->toRegex($pattern))->render();
+
+            if (preg_match('/\W([imsxUXJ]*)$/', $regex, $match)) {
+                $regexps[] = '(?'.$match[1].':'.substr($regex, 1, -strlen($match[0])).')';
+            } else {
+                $this->noMatchRegexps[] = $regex;
+            }
+        }
+        if ($regexps) {
+            $this->noMatchRegexps[] = Regex::BOUNDARY.implode('|', $regexps).Regex::BOUNDARY;
         }
 
         parent::__construct($iterator);
diff --git a/src/Symfony/Component/Finder/Tests/Iterator/MultiplePcreFilterIteratorTest.php b/src/Symfony/Component/Finder/Tests/Iterator/MultiplePcreFilterIteratorTest.php
index 89d8edb..3b7a9f0 100644
--- a/src/Symfony/Component/Finder/Tests/Iterator/MultiplePcreFilterIteratorTest.php
+++ b/src/Symfony/Component/Finder/Tests/Iterator/MultiplePcreFilterIteratorTest.php
@@ -31,7 +31,7 @@ class MultiplePcreFilterIteratorTest extends \PHPUnit_Framework_TestCase
             array(' foo ', false, '" " is not a valid delimiter'),
             array('\\foo\\', false, '"\\" is not a valid delimiter'),
             array('afooa', false, '"a" is not a valid delimiter'),
-            array('//', false, 'the pattern should contain at least 1 character'),
+            array('//', true, 'valid empty regex'),
             array('/a/', true, 'valid regex'),
             array('/foo/', true, 'valid regex'),
             array('/foo/i', true, 'valid regex with a single modifier'),
```

**Comment by nicolas-grekas** [comment]
once there is more than one rx, merging regexps is a bit faster, path added to the PR
one tests is failing on hhvm
status: needs work

File: src/Symfony/Component/Finder/Expression/Regex.php [hunk 4]
```diff
@@ -61,7 +61,7 @@ class Regex implements ValueInterface
      */
     public static function create($expr)
     {
-        if (preg_match('/^(.{3,}?)([imsxuADU]*)$/', $expr, $m)) {
+        if (preg_match('/^(.{2,}?)([imsxuADU]*)$/', $expr, $m)) {
```
Comment (stof) [inline src/Symfony/Component/Finder/Expression/Regex.php]: what is the reason for this change ?

File: src/Symfony/Component/Finder/Iterator/ExcludeDirectoryFilterIterator.php [hunk 5]
```diff
@@ -29,10 +31,18 @@ class ExcludeDirectoryFilterIterator extends FilterIterator implements \Recursiv
...
         foreach ($directories as $directory) {
-            $this->patterns[] = '#(^|/)'.preg_quote($directory, '#').'(/|$)#';
+            if (!$this->isRecursive || false !== strpos($directory, '/')) {
+                $patterns[] = preg_quote($directory, '#');
+            } else {
+                $this->excludedDirs[$directory] = true;
+            }
+        }
+        if ($patterns) {
+            $this->excludedPattern = '#(^|/)('.implode('|', $patterns).')(/|$)#';
```
Comment (stof) [inline src/Symfony/Component/Finder/Iterator/ExcludeDirectoryFilterIterator.php]: I suggest using non-capturing groups here. It will make the regex matching more efficient (the regex engine will not need to capture the groups)

File: src/Symfony/Component/Finder/Iterator/MultiplePcreFilterIterator.php [hunk 6]
```diff
@@ -65,4 +59,28 @@ protected function isRegex($str)
...
+        }
+        if (1 < count($rxs)) {
+            $regexps[] = Regex::BOUNDARY.implode('|', $rxs).Regex::BOUNDARY;
+        } elseif ($rxs) {
+            $rxs = explode(':', substr($rxs[0], 2, -1), 2);
+            $regexps[] = Regex::BOUNDARY.$rxs[1].Regex::BOUNDARY.$rxs[0];
+        }
+
+        return $regexps;
+    }
```
Comment (stof) [inline src/Symfony/Component/Finder/Iterator/MultiplePcreFilterIterator.php]: this logic will break regexes using complex features, for instance ``(?:^|/)(\w)\w+\1(?:/|$)`` to match directories starting and ending with the same char in the PathFilter.
combining regexes together is impossible if you don't have any knowledge about these regexes.

I suggest you to revert this change for now, to make the PR mergeable.

I have an idea to optimize the existing filters extending this class though, by using some knowledge about the way regexes are built currently (because the user input may not be a regex but something we are turning into a regex, and so we can optimize these cases). I will work on it this evening if I can.

File: src/Symfony/Component/Finder/Expression/Regex.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Finder/Expression/Regex.php]: the regex class rejects empty expressions `//`, but it should not, they are perfectly valid

File: src/Symfony/Component/Finder/Iterator/MultiplePcreFilterIterator.php (hunk 6 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Finder/Iterator/MultiplePcreFilterIterator.php]: OK, removed. Just in case and for history, I'll open an other PR with this patch, with a fix for excluding unmergeable regexps.

**Comment by stof** [comment]
:+1: on this one

**Comment by nicolas-grekas** [comment]
hhvm fixed, good to merge
Status: reviewed

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.
